### PR TITLE
ENH: add kwargs `rename` to `Styler.format_index` for overwriting index labels

### PR DESCRIPTION
**Pull request by attack68**
In methods such as `DataFrame.to_latex` and `DataFrame.to_html` the `header` keyword arg was implemented to provide aliases for column headers, instead of using the underlying DataFrame labels.

Currently `Styler.format_index` allows adding a generic formatter to change the display of index labels but this is not an easy way to simply provide a list of string aliases to use instead. PR adds this function, which will simplify the transition of the above methods to Styler implementation.

### CODE REVIEW

**Comment by attack68** [comment]
> ok so is the plan to deprecate `header`? why are you adding this rather than we just use the given labels?

3 small arguments for use case:

First, progressing `Styler` as a visualisation tool allowing full customisation of the display, independent of the data and label values. This is already possible where you can now apply a formatting function (1 variable input -> str) to indexes, which is very useful, for example, when formatting dates. But a generic function cannot just overwrite the (visible) labels with user specified values, but this `aliases` arg will allow that.

Second, the existing IO methods have this `headers` arg, which allows overwriting column labels with aliases. Personally I think its not sensible to try and squish all the `Styler` functionality (via chaining) into a single `DataFrame.to_html` and `to_latex` call, but the path of least resistance (community approvals wise) seems to be to not take away functionality from existing kwargs. I.e we would likely keep `headers=xx` and use the styler implementation just refactor as a call to `format_index(aliases=xx)`.

Finally, in a more specialised scenario you might want a formatter that uses a loop index, or combines all level values simulatneously to determine an alias, which you then pass directly here. Note that the indexing of the data can still be maintained underneath via the MultiIndex.

```
multiindex = MultiIndex.from_product([["Mac", "Windows"],  ["2GB", "4GB"]])
def relabel(os, ram, loop):
    return f"{loop}_{os}_{ram}"
aliases = [relabel(idx[0], idx[1], i) for i, idx in enumerate(multiindex)]
styler.hide(level=0).format_index(aliases=aliases)

      level 1           data
0_Mac_2GB                  x
1_Mac_4GB                  x
2_Windows_2GB              x
3_Windows_4GB              x
```

**Comment by attack68** [comment]
Screen Shot 2022-01-20 at 19 01 25
Screen Shot 2022-01-20 at 19 01 40
Screen Shot 2022-01-20 at 19 01 55
Screen Shot 2022-01-20 at 19 02 02
Screen Shot 2022-01-20 at 19 02 16

**Review by jreback — CHANGES_REQUESTED** [review]
not objecting to the need to do this, i get it. but we don't use `alias` *anywhere*; why isn't this *just* `.rename()`?

**Comment by attack68** [comment]
> not objecting to the need to do this, i get it. but we don't use `alias` _anywhere_; why isn't this _just_ `.rename()`?

`rename` loosely suggests the data is modified, when it is really just displayed. Therefore, included within the `format_index` method.

Didn't really feel this warranted its own method when all other index formatting is done within `format_index`.
(and its not possible to  combine the two so there was that also)

But im not tied to anything; you tell me which you prefer and ill make changes...

**Comment by attack68** [comment]
@jreback what about a different kwarg other that `alias`, how about `names` or `labels`?

File: pandas/io/formats/style_render.py [hunk 1]
```diff
@@ -1203,6 +1207,17 @@ def format_index(
             Convert string patterns containing https://, http://, ftp:// or www. to
             HTML <a> tags as clickable URL hyperlinks if "html", or LaTeX \href
             commands if "latex".
+        rename : list of str, list of list of str
```
Comment (jreback) [inline pandas/io/formats/style_render.py:1210]: umm the user can simply `.rename_axis` on the frame no? why adding api here that is duplicative

Comment (attack68) [reply pandas/io/formats/style_render.py:1210]: my design objective with styler has been to customise the display of tables without altering the underlying dataframe object (which is of course possible to get around many of the styler functions).

you can lose the power of indexing if you relabel for the purpose of printout.

additionally, methods like `DataFrame.to_html` and `DataFrame.to_latex` use a `headers` argument which does this relabelling. So in order to replicate that behaviour with the styler code need this pr.

Comment (attack68) [reply pandas/io/formats/style_render.py:1210]: here's a good example:
Screenshot 2022-03-07 at 18 14 58

**Comment by attack68** [comment]
@jreback feels like this one has stalled and getting push back anyway. If poss please take one last look since it factors into the PRs where I convert the `DataFrame.to_latex` and `DataFrame.to_html` to use the Styler implementation. And if this one is closed then I have to come up with another way of refactoring them. thx

**Comment by jreback** [comment]
@attack68 circling back around on this.

**Comment by attack68** [comment]
In order to be able to completely re-write `DataFrame.to_latex` with the `Styler` implementation, this is the last piece of functionality needed on the Styler side. Then issues being faced by users, such as [stackoverflow q](https://stackoverflow.com/questions/72780890/handle-futurewarning-by-converting-pandas-dataframe-table-to-latex-with-function/72793009#72793009), can be automatically addressed.

**Comment by jreback** [comment]
@attack68 i am sympathetic here as i know you want to make this fully compatible with the existing formaters. The issues i have here are:
- adding yet another keyword to the api, this bloating the api even more.
- we are merging the concerns here. formatting output data and replacing the data, which we *already* have a good way of doing via `.rename_axis` / `.set_axis`. This leads to more than one way to do something. 

I agree that this could be viewed as on the line, but its much nicer to keep this api nice and tight and use the existing fluent api to solve this problem.

I see that we *already* have styler methods to do some index manipulation e.g. https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.io.formats.style.Styler.format_index.html?highlight=styler (`.apply_index`, `format_index`). I guess ok to add a `.rename_index` as a method here might be ok. (this preserves the fluent api and doesn't add yet another keyword).

**Comment by attack68** [comment]
@jreback, OK, yes I see your point, `rename_index` could also serve an additional purpose, which I will explain in the docs write up. Will look to do that instead.

**Pull request closed without merging**